### ISSUE TITLE
no-invalid-this fails on class properties

### ISSUE THREAD

**Issue by j-f1** [eslint/eslint#9075]
**Tell us about your environment**
I’m using TypeScript, but this error is reproducible in Babel/JS, too.

* **ESLint Version:** v4.4.0
* **Node Version:** v7.9.0
* **npm Version:** 5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?**
TypeScript

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yaml
parser: typescript-eslint-parser
plugins:
  - typescript
  - react
extends:
  - prettier
  - prettier/react

rules:
  ##########
  # CUSTOM #
  ##########
  insecure-random: error

  ###########
  # PLUGINS #
  ###########
  typescript/interface-name-prefix:
    - error
    - always
  typescript/explicit-member-accessibility: error
  typescript/no-unused-vars: error
  # blocked by https://github.com/nzakas/eslint-plugin-typescript/pull/23
  # typescript/member-ordering: error

  ###########
  # BUILTIN #
  ###########
  curly: error
  # bug with JSDoc
  # indent:
  #   - error
  #   - 2
  #   - SwitchCase: 1

  array-bracket-spacing:
    - warn
    - never
  object-curly-spacing:
    - warn
    - always
  no-new-wrappers: error
  no-redeclare:
    - error
    - builtinGlobals: true
  no-eval: error
  no-invalid-this: error
  no-sync: error
  no-trailing-spaces:
    - error
    # why?
    - ignoreComments: true
  no-unused-expressions: error
  no-use-before-define:
    - error
    - functions: false
      variables: false
  no-var: error
  prefer-const: error
  no-extra-semi: error
  semi:
    - error
    - never
  comma-dangle:
    - error
    - always-multiline
  eqeqeq:
    - error
    - smart
  space-before-blocks: error
  keyword-spacing: error
  space-infix-ops: error
  semi-spacing: error
  comma-spacing: error

  ###########
  # SPECIAL #
  ###########
  no-restricted-syntax:
    - error
    # no-default-export
    - selector: ExportDefaultDeclaration
      message: Use of default exports is forbidden

parserOptions:
  sourceType: module
  ecmaFeatures:
    jsx: true
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```ts
class UpdateStore {
  // ..
  private onAutoUpdaterError = (error: Error) => {
    // If we get an error during any stage of the update process we'll
    this.status = UpdateStatus.UpdateNotAvailable
    this.emitError(error)
  }
  // ...
}
```

**What did you expect to happen?**
Lint passes

**What actually happened? Please include the actual, raw output from ESLint.**

```
/path/to/desktop/app/src/ui/app-error.tsx
   63:26  error  Unexpected 'this'  no-invalid-this
```

---

This appears to be because the [`isDefaultThisBinding` implementation](https://github.com/eslint/eslint/blob/a113cd3bf831078c3d7d417aa7ab768b2fbd0fe4/lib/ast-utils.js#L603) is missing a case for `ClassProperty`. It is possible that the fix could be as simple as inserting `case "ClassProperty":` near [this line](https://github.com/eslint/eslint/blob/a113cd3bf831078c3d7d417aa7ab768b2fbd0fe4/lib/ast-utils.js#L647).

**Comment by not-an-aardvark** [comment]
Closing as a duplicate of https://github.com/eslint/eslint/issues/8271.
